File: day03/assignments/__epyccel__/__pyccel__/reduction_tab_x45ty7lcn9h9_x45ty7lcn9h9.c
#include "reduction_tab_x45ty7lcn9h9_x45ty7lcn9h9.h"
#include <stdlib.h>
#include <stdint.h>
#include "ndarrays.h"
#include "omp.h"


/*........................................*/
int32_t threads_num(void)
{
    /*complete this function for getting threads*/
    return omp_get_num_threads();
}
/*........................................*/
/*........................................*/
void initialize_table(t_ndarray tab, int64_t nmolec, int64_t n, int64_t nmol)
{
    int64_t k;
    int64_t j;
    int64_t i;
    /*Initialisation du tableau*/
    /*First-touch pour garantir un fonctionnement optimal sur les systemes NUMA*/
    #pragma omp parallel
    {
        #pragma omp for
        for (k = 0; k < nmolec; k += 1)
        {
            for (j = 0; j < n; j += 1)
            {
                for (i = 0; i < nmol; i += 1)
                {
                    GET_ELEMENT(tab, nd_double, (int64_t)i, (int64_t)j, (int64_t)k) = i + j + k;
                }
            }
        }
    }
}
/*........................................*/
/*........................................*/
void reduction(t_ndarray tab, t_ndarray tab1, t_ndarray tab2, int64_t nmolec, int64_t nmol, int64_t n)
{
    int64_t k;
    int64_t j;
    int64_t i;
    int64_t i_0001;
    #pragma omp parallel
    {
        #pragma omp for
        for (k = 0; k < nmolec; k += 1)
        {
            for (i_0001 = 0; i_0001 < nmol; i_0001 += 1)
            {
                GET_ELEMENT(tab1, nd_double, (int64_t)i_0001) = 0;
            }
            for (j = 0; j < n; j += 1)
            {
                for (i = 0; i < nmol; i += 1)
                {
                    GET_ELEMENT(tab1, nd_double, (int64_t)i) = GET_ELEMENT(tab1, nd_double, (int64_t)i) + GET_ELEMENT(tab, nd_double, (int64_t)i, (int64_t)j, (int64_t)k);
                }
            }
            for (i_0001 = 0; i_0001 < nmol; i_0001 += 1)
            {
                GET_ELEMENT(tab2, nd_double, (int64_t)i_0001) = GET_ELEMENT(tab2, nd_double, (int64_t)i_0001) + 2 * GET_ELEMENT(tab1, nd_double, (int64_t)i_0001);
            }
        }
    }
}
/*........................................*/

File: day03/assignments/__epyccel__/reduction_tab_x45ty7lcn9h9_x45ty7lcn9h9.py

def threads_num():
#     complete this function for getting threads 
    from pyccel.stdlib.internal.openmp import omp_get_num_threads 
    return omp_get_num_threads()

def initialize_table(tab:'double[:,:,:]', nmolec:'int', n:'int', nmol:'int'):
    # Initialisation du tableau
    # First-touch pour garantir un fonctionnement optimal sur les systemes NUMA
    
    #$ omp parallel
    #$ omp for 
    for k in range(nmolec):
        for j in range(n):
            for i in range(nmol):
                tab[ i,j,k ] = i + j + k
                
    #$ omp end parallel

def reduction(tab:'double[:,:,:]', tab1:'double[:]', tab2:'double[:]', nmolec:'int', nmol:'int', n:'int'):
    
    #$ omp parallel
    #$ omp for 
    for k in range(nmolec):
        tab1[:nmol] = 0
        for j in range(n):
            for i in range(nmol):
                tab1[i] = tab1[i] + tab[i,j,k]
        tab2[:nmol] = tab2[:nmol] + 2 * tab1[:nmol]
        
    #$ omp end parallel
